injection/sharedmain/main.go

```diff
@@ -143,102 +166,180 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
+			logger.Fatalw("Version check failed", zap.Error(err))
+		}
 
-	// Create ConfigMaps watcher with optional label-based filter.
-	var cmLabelReqs []labels.Requirement
-	if cmLabel := system.ResourceLabel(); cmLabel != "" {
-		req, err := configmap.FilterConfigByLabelExists(cmLabel)
-		if err != nil {
-			logger.With(zap.Error(err)).Fatalf("Failed to generate requirement for label %q")
+		// Create ConfigMaps watcher with optional label-based filter.
+		var cmLabelReqs []labels.Requirement
+		if cmLabel := system.ResourceLabel(); cmLabel != "" {
+			req, err := configmap.FilterConfigByLabelExists(cmLabel)
+			if err != nil {
+				logger.With(zap.Error(err)).Fatalf("Failed to generate requirement for label %q")
+			}
+			logger.Infof("Setting up ConfigMap watcher with label selector %q", req)
+			cmLabelReqs = append(cmLabelReqs, *req)
 		}
-		logger.Infof("Setting up ConfigMap watcher with label selector %q", req)
-		cmLabelReqs = append(cmLabelReqs, *req)
-	}
-	// TODO(mattmoor): This should itself take a context and be injection-based.
-	cmw := configmap.NewInformedWatcher(kc, system.Namespace(), cmLabelReqs...)
-
-	// Based on the reconcilers we have linked, build up the set of controllers to run.
-	controllers := make([]*controller.Impl, 0, len(ctors))
-	webhooks := make([]interface{}, 0)
-	for _, cf := range ctors {
-		ctrl := cf(ctx, cmw)
-		controllers = append(controllers, ctrl)
-
-		// Build a list of any reconcilers that implement webhook.AdmissionController
-		switch c := ctrl.Reconciler.(type) {
-		case webhook.AdmissionController, webhook.ConversionController:
-			webhooks = append(webhooks, c)
+		// TODO(mattmoor): This should itself take a context and be injection-based.
+		cmw := configmap.NewInformedWatcher(kc, system.Namespace(), cmLabelReqs...)
+
+		// Based on the reconcilers we have linked, build up the set of controllers to run.
+		controllers := make([]*controller.Impl, 0, len(ctors))
+		webhooks := make([]interface{}, 0)
+		for _, cf := range ctors {
+			ctrl := cf(ctx, cmw)
+			controllers = append(controllers, ctrl)
+
+			// Build a list of any reconcilers that implement webhook.AdmissionController
+			switch c := ctrl.Reconciler.(type) {
+			case webhook.AdmissionController, webhook.ConversionController:
+				webhooks = append(webhooks, c)
+			}
 		}
-	}
 
-	profilingHandler := profiling.NewHandler(logger, false)
+		profilingHandler := profiling.NewHandler(logger, false)
 
-	// Watch the logging config map and dynamically update logging levels.
-	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(logging.ConfigMapName(),
-		metav1.GetOptions{}); err == nil {
-		cmw.Watch(logging.ConfigMapName(), logging.UpdateLevelFromConfigMap(logger, atomicLevel, component))
-	} else if !apierrors.IsNotFound(err) {
-		logger.With(zap.Error(err)).Fatalf("Error reading ConfigMap %q", logging.ConfigMapName())
-	}
+		// Watch the logging config map and dynamically update logging levels.
```

Comment:
Since these are configured as part of the leader-elected code, I think that there is no notion where these configs apply when the controller is not the leader. If the controller is running and not the leader it is waiting to be the leader. Note, there is an exceptional case where the leader election lock may be lost but the process still running due to deadlock which will be addressed by #1048, but I don't think it changes the equation, and I think it would be fine to have logging and observability config watches stay within the leader-elected section.